File: video_merger.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

var videoExtensions = []string{".mp4", ".avi", ".mov", ".mkv", ".flv", ".wmv"}

func isVideoFile(filename string) bool {
	if strings.Count(filename, ".") != 1 {
		return false
	}
	ext := strings.ToLower(filepath.Ext(filename))
	for _, v := range videoExtensions {
		if ext == v {
			return true
		}
	}
	return false
}

// 自然排序实现
type naturalStrings []string

func (ns naturalStrings) Len() int           { return len(ns) }
func (ns naturalStrings) Swap(i, j int)      { ns[i], ns[j] = ns[j], ns[i] }
func (ns naturalStrings) Less(i, j int) bool { return naturalLess(ns[i], ns[j]) }

var numRegexp = regexp.MustCompile(`(\d+)`)

func naturalLess(a, b string) bool {
	aa := numRegexp.FindAllStringIndex(a, -1)
	bb := numRegexp.FindAllStringIndex(b, -1)
	ai, bi := 0, 0
	for ai < len(aa) && bi < len(bb) {
		// 比较前缀
		if a[:aa[ai][0]] != b[:bb[bi][0]] {
			return a[:aa[ai][0]] < b[:bb[bi][0]]
		}
		// 比较数字
		numA, _ := strconv.Atoi(a[aa[ai][0]:aa[ai][1]])
		numB, _ := strconv.Atoi(b[bb[bi][0]:bb[bi][1]])
		if numA != numB {
			return numA < numB
		}
		// 截断前缀和数字，继续比较
		a = a[aa[ai][1]:]
		b = b[bb[bi][1]:]
		aa = numRegexp.FindAllStringIndex(a, -1)
		bb = numRegexp.FindAllStringIndex(b, -1)
		ai, bi = 0, 0
	}
	return a < b
}

func findVideoFiles(dir string) ([]string, error) {
	files, err := ioutil.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	var videoFiles []string
	for _, file := range files {
		if !file.IsDir() && isVideoFile(file.Name()) {
			videoFiles = append(videoFiles, filepath.Join(dir, file.Name()))
		}
	}
	sort.Sort(naturalStrings(videoFiles))
	return videoFiles, nil
}

func createFileList(videoFiles []string, listPath string) error {
	f, err := os.Create(listPath)
	if err != nil {
		return err
	}
	defer f.Close()
	w := bufio.NewWriter(f)
	for _, file := range videoFiles {
		fmt.Printf("Merging: %s\n", file)
		_, err := w.WriteString(fmt.Sprintf("file '%s'\n", file))
		if err != nil {
			return err
		}
	}
	return w.Flush()
}

func getFFmpegPath() string {
	// 优先查找当前目录下的 ffmpeg.exe（Windows）或 ffmpeg（其他平台）
	ffmpegName := "ffmpeg"
	if os.PathSeparator == '\\' {
		ffmpegName = "ffmpeg.exe"
	}
	cwd, err := os.Getwd()
	if err == nil {
		ffmpegPath := filepath.Join(cwd, ffmpegName)
		if _, err := os.Stat(ffmpegPath); err == nil {
			return ffmpegPath
		}
	}
	// 否则用系统 PATH
	return ffmpegName
}

func mergeVideos(videoFiles []string, output string) error {
	if len(videoFiles) == 0 {
		return fmt.Errorf("no video files to merge")
	}
	listPath := "video_list.txt"
	err := createFileList(videoFiles, listPath)
	if err != nil {
		return err
	}
	defer os.Remove(listPath)

	cmd := exec.Command(getFFmpegPath(), "-f", "concat", "-safe", "0", "-i", listPath, "-c:v", "libx264", "-c:a", "aac", "-strict", "-2", output)
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return err
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return err
	}
	if err := cmd.Start(); err != nil {
		return err
	}
	// 实时输出 ffmpeg 的进度
	go io.Copy(os.Stdout, stdout)
	scanner := bufio.NewScanner(stderr)
	for scanner.Scan() {
		line := scanner.Text()
		fmt.Println(line)
	}
	if err := scanner.Err(); err != nil {
		return err
	}
	return cmd.Wait()
}

func main() {
	dir := flag.String("dir", "", "Directory containing video files to merge.")
	output := flag.String("output", "", "Filename for the merged output video.")
	flag.Parse()

	if *dir == "" || *output == "" {
		fmt.Println("Usage: video_merger_go --dir <video_dir> --output <output_file>")
		os.Exit(1)
	}

	videoFiles, err := findVideoFiles(*dir)
	if err != nil {
		fmt.Printf("Error reading directory: %v\n", err)
		os.Exit(1)
	}
	if len(videoFiles) == 0 {
		fmt.Printf("No video files found in directory '%s'.\n", *dir)
		os.Exit(1)
	}

	fmt.Printf("Found %d video files to merge:\n", len(videoFiles))
	for _, vf := range videoFiles {
		fmt.Println("  -", vf)
	}

	err = mergeVideos(videoFiles, *output)
	if err != nil {
		fmt.Printf("Failed to merge videos: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("\nSuccessfully merged videos into '%s'\n", *output)
}

File: video_merger.py
import os
import argparse
from moviepy.editor import VideoFileClip, concatenate_videoclips
from moviepy.config import change_settings

def find_video_files(directory):
    """Finds video files in the specified directory."""
    video_extensions = ['.mp4', '.avi', '.mov', '.mkv', '.flv', '.wmv']
    video_files = []
    if not os.path.isdir(directory):
        print(f"Error: Directory '{directory}' not found.")
        return None
    for item in os.listdir(directory):
        if os.pathisfile(os.path.join(directory, item)):
            _, ext = os.path.splitext(item)
            if ext.lower() in video_extensions:
                video_files.append(os.path.join(directory, item))
    return sorted(video_files) # Sort to ensure consistent order

def merge_videos(video_files, output_filename):
    """Merges a list of video files into a single output file."""
    if not video_files:
        print("No video files found to merge.")
        return False

    print(f"Found {len(video_files)} video files to merge:")
    for vf in video_files:
        print(f"  - {vf}")

    clips = []
    try:
        for video_file in video_files:
            try:
                clip = VideoFileClip(video_file)
                clips.append(clip)
            except Exception as e:
                print(f"Warning: Could not load video file '{video_file}': {e}")
                print("This file will be skipped.")

        if not clips:
            print("No valid video clips could be loaded. Merging aborted.")
            return False

        final_clip = concatenate_videoclips(clips, method="compose")
        final_clip.write_videofile(output_filename, codec="libx264", audio_codec="aac")

        # Close all clips
        for clip in clips:
            clip.close()
        if hasattr(final_clip, 'close') and callable(final_clip.close):
            final_clip.close()

        print(f"\nSuccessfully merged videos into '{output_filename}'")
        return True
    except ImportError:
        print("MoviePy library not found. Please install it: pip install moviepy")
        return False
    except OSError as e:
        if "ffmpeg" in str(e).lower() or "imageio" in str(e).lower() :
            print(f"Error: FFmpeg might be missing or not configured correctly for MoviePy. {e}")
            print("Please ensure FFmpeg is installed and accessible in your system's PATH.")
            print("You can download it from https://ffmpeg.org/download.html")
            print("If FFmpeg is installed, you might need to configure MoviePy by editing moviepy/config_defaults.py or setting FFMPEG_BINARY environment variable.")
            # Attempt to guide user for manual config if auto-detection fails
            # This is a common issue with moviepy
            try:
                # Example of how one might try to set it if they know the path
                # change_settings({"FFMPEG_BINARY": "/path/to/your/ffmpeg"}) # Or "ffmpeg.exe" on Windows
                print("\nIf you have ffmpeg installed but not in PATH, you can try setting the FFMPEG_BINARY path in moviepy's config_defaults.py")
                print("Or, set the FFMPEG_BINARY environment variable pointing to your ffmpeg executable.")
            except Exception as config_e:
                print(f"Could not apply dynamic FFMPEG_BINARY setting: {config_e}")

        else:
            print(f"An OS error occurred: {e}")
        return False
    except Exception as e:
        print(f"An unexpected error occurred during merging: {e}")
        return False

def main():
    parser = argparse.ArgumentParser(description="Merge video files from a directory.")
    parser.add_argument("--dir", type=str, required=True, help="Directory containing video files to merge.")
    parser.add_argument("--output", type=str, required=True, help="Filename for the merged output video.")

    args = parser.parse_args()

    video_files = find_video_files(args.dir)

    if video_files is None: # Directory not found
        return

    if not video_files:
        print(f"No video files found in directory '{args.dir}'.")
        return

    merge_videos(video_files, args.output)

if __name__ == "__main__":
    main()
